pkg/kubectl/cmd/set/set_volume.go

```diff
@@ -0,0 +1,980 @@
+	  # List volumes defined on all deployment in the current project
+	  kubectl set volume deployment --all
+
+	  # Add a new empty dir volume to deployment 'registry' mounted under
+	  # /var/lib/registry
+	  kubectl set volume deployment/registry --add --mount-path=/var/lib/registry
+
+	  # Use an existing persistent volume claim (pvc) to overwrite an existing volume 'v1'
+	  kubectl set volume deployment/registry --add --name=v1 -t pvc --claim-name=pvc1 --overwrite
+
+	  # Remove volume 'v1' from deployment 'registry'
+	  kubectl set volume deployment/registry --remove --name=v1
+
+	  # Create a new persistent volume claim that overwrites an existing volume 'v1'
+	  kubectl set volume deployment/registry --add --name=v1 -t pvc --claim-size=1G --overwrite
+
+	  # Change the mount point for volume 'v1' to /data
+	  kubectl set volume deployment/registry --add --name=v1 -m /data --overwrite
+
+	  # Modify the deployment by removing volume mount "v1" from container "c1"
+	  # (and by removing the volume "v1" if no other containers have volume mounts that reference it)
+	  kubectl set volume deployment/registry --remove --name=v1 --containers=c1
+
+	  # Add new volume based on a more complex volume source (Git repo, AWS EBS, GCE PD,
+	  # Ceph, Gluster, NFS, ISCSI, ...)
+	  kubectl set volume deployment/registry --add -m /repo --source=<json-string>`)
+)
+
+type VolumeOptions struct {
+	DefaultNamespace       string
+	EnforceNamespace       bool
+	Out                    io.Writer
+	Err                    io.Writer
+	Mapper                 meta.RESTMapper
+	RESTClientFactory      func(mapping *meta.RESTMapping) (resource.RESTClient, error)
+	UpdatePodSpecForObject func(obj runtime.Object, fn func(*v1.PodSpec) error) (bool, error)
+	Client                 kcoreclient.PersistentVolumeClaimsGetter
+	Encoder                runtime.Encoder
+	Cmd                    *cobra.Command
+
+	// Resource selection
+	resource.FilenameOptions
+	Selector string
+	All      bool
+
+	// Operations
+	Add    bool
+	Remove bool
+	List   bool
+
+	// Common optional params
+	Name        string
+	Containers  string
+	Confirm     bool
+	Local       bool
+	DryRun      bool
+	Output      string
+	PrintObject func(cmd *cobra.Command, isLocal bool, mapper meta.RESTMapper, obj runtime.Object, out io.Writer) error
+
+	// Add op params
+	AddOpts *AddVolumeOptions
+}
+
+type AddVolumeOptions struct {
+	Type          string
+	MountPath     string
+	SubPath       string
+	DefaultMode   string
+	Overwrite     bool
+	Path          string
+	ConfigMapName string
+	SecretName    string
+	Source        string
+
+	CreateClaim bool
+	ClaimName   string
+	ClaimSize   string
+	ClaimMode   string
+	ClaimClass  string
+
+	TypeChanged bool
+}
+
+func NewCmdVolume(f cmdutil.Factory, out, errOut io.Writer) *cobra.Command {
+	addOpts := &AddVolumeOptions{}
+	opts := &VolumeOptions{AddOpts: addOpts}
+	cmd := &cobra.Command{
+		Use:     "volumes RESOURCE/NAME --add|--remove|--list",
```

Comment:
The `--remove` functionality will likely end up in its own [`kubectl unset`](https://github.com/kubernetes/kubernetes/issues/39775) command and `--list` in [`kubectl view`](https://github.com/kubernetes/kubernetes/issues/29679). That being said,I would suggest making `--add` the default here and remove all of these flags. 

cc: @kubernetes/sig-cli-pr-reviews